File: web/dart/blocks/clause.dart
/*
 * NetTango
 * Copyright (c) 2019 Michael S. Horn, Uri Wilensky, and Corey Brady
 *
 * Northwestern University
 * 2120 Campus Drive
 * Evanston, IL 60613
 * http://tidal.northwestern.edu
 * http://ccl.northwestern.edu

 * This project was funded in part by the National Science Foundation.
 * Any opinions, findings and conclusions or recommendations expressed in this
 * material are those of the author(s) and do not necessarily reflect the views
 * of the National Science Foundation (NSF).
 */
part of NetTango;

class Clause extends BlockCollection {

  final Block owner;
  final int clauseIndex;

  bool isDragOver = false;
  bool isDragHeaderOver = false;

  Clause(this.owner, {this.clauseIndex = null}) {}

  static Clause fromJSON(CodeWorkspace workspace, Block owner, Map json, int clauseIndex) {
    Clause clause = new Clause(owner, clauseIndex: clauseIndex);
    if (json["children"] is List) {
      clause.blocks = BlockCollection.fromJSON(workspace, json["children"]);
    }
    return clause;
  }

  DivElement draw(DragImage dragImage, DivElement headerDiv) {
    _div = new DivElement();
    _div.classes.add("nt-clause");

    final headerDropzone = Dropzone(headerDiv, acceptor: owner.workspace.blockAcceptor);
    headerDropzone.onDrop.listen(drop);
    headerDropzone.onDragEnter.listen( (e) => isDragHeaderOver = true );
    headerDropzone.onDragLeave.listen( (e) => isDragHeaderOver = false );

    final dropzone = Dropzone(_div, acceptor: owner.workspace.blockAcceptor);
    dropzone.onDrop.listen(drop);
    dropzone.onDragEnter.listen( (e) => isDragOver = true );
    dropzone.onDragLeave.listen( (e) => isDragOver = false );

    if (blocks.isEmpty) {
      setEmpty();
      return _div;
    }

    for (int i = 0; i < blocks.length; i++) {
      Block block = blocks[i];
      final siblings = blocks.skip(i + 1);
      final dragData = BlockDragData.blockOwned(owner._dragData.chainIndex, i, owner.instanceId, siblings, clauseIndex: clauseIndex);
      block.draw(dragImage, dragData);
    }

    BlockCollection.appendBlocks(_div, blocks, "nt-block-clause");

    return _div;
  }

  void setEmpty() {
    _div.classes.add("nt-clause-empty");
    _div.append(Notch.drawClause(false, this));

    final dropElement = new DivElement() .. className = "nt-clause-drop";
    _div.append(dropElement);

    _div.append(Notch.drawClause(true,  this));
  }

  void setNonEmpty() {
    _div.classes.remove("nt-clause-empty");
  }

  void resetOwned() {
    for (int i = 0; i < blocks.length; i++) {
      Block block = blocks[i];
      block._dragData.resetBlockOwned(owner._dragData.chainIndex, i, owner.instanceId, blocks.skip(i + 1), clauseIndex: clauseIndex);
      block.resetOwnedBlocksDragData();
    }
  }

  void redrawBlocks() {
    _div.innerHtml = "";

    if (blocks.isEmpty) {
      setEmpty();
      return;
    }

    setNonEmpty();

    for (int i = 0; i < blocks.length; i++) {
      Block block = blocks[i];
      block._dragData.resetBlockOwned(owner._dragData.chainIndex, i, owner.instanceId, blocks.skip(i + 1), clauseIndex: clauseIndex);
      block.resetOwnedBlocksDragData();
    }

    BlockCollection.appendBlocks(_div, blocks, "nt-block-clause");
  }

  bool updateDragOver() {
    _div.classes.remove("nt-block-clause-drag-over");
    if (blocks.isNotEmpty) { blocks[0]._blockDiv.classes.remove("nt-block-clause-drag-over"); }
    bool isHighlightHandled = false;
    for (Block block in blocks) {
      final blockResult = block.updateDragOver();
      isHighlightHandled = isHighlightHandled || blockResult;
    }
    if ((isDragOver || isDragHeaderOver) && !isHighlightHandled) {
      isHighlightHandled = true;
      if (blocks.isEmpty) {
        _div.classes.add("nt-block-clause-drag-over");
      } else {
        blocks[0]._blockDiv.classes.add("nt-block-clause-drag-over");
      }
    }
    return isHighlightHandled;
  }

  void clearDragOver() {
    _div.classes.remove("nt-block-clause-drag-over");
    if (blocks.isNotEmpty) { blocks[0]._blockDiv.classes.remove("nt-block-clause-drag-over"); }
    isDragOver = false;
    isDragHeaderOver = false;
    for (Block block in blocks) {
      block.clearDragOver();
    }
  }

  void drop(DropzoneEvent event) {
    DragAcceptor.wasHandled = true;

    final newBlocks = owner.workspace.consumeDraggingBlocks();

    insertBlocks(0, newBlocks);
    _div.classes.remove("nt-clause-empty");

    Block changedBlock = newBlocks.elementAt(0);
    owner.workspace.programChanged(new BlockChangedEvent(changedBlock));
  }

  void insertBlocks(int blockIndex, Iterable<Block> newBlocks) {
    blocks.insertAll(blockIndex, newBlocks);
    redrawBlocks();
  }

  Iterable<Block> removeBlocks(int blockIndex) {
    final removed = blocks.skip(blockIndex);
    blocks = blocks.take(blockIndex).toList();
    redrawBlocks();
    return removed;
  }

}

File: src/blocks/clause.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import interact from "interactjs"

import { StringUtils } from "../utils/string-utils"
import { Arrow } from "./baubles/arrow"
import { Notch } from "./baubles/notch"
import { Block } from "./block-instance"
import { BlockCollection } from "./block-collection"
import { ClauseAcceptor } from "./drag-drop/clause-acceptor"
import { ActiveDragData } from "./drag-drop/drag-data/active-drag-data"
import { ClauseDragData } from "./drag-drop/drag-data/clause-drag-data"
import { DragManager } from "./drag-drop/drag-manager"
import { BlockChangedEvent } from "./program-changed-event"
import { ClauseInput, ClauseInstanceInput } from "../types/types"
import { BlockRules } from "./block-rules"

class Clause extends BlockCollection {

  readonly def: ClauseInput
  readonly c: ClauseInstanceInput
  readonly owner: Block
  readonly clauseIndex: number

  divider: HTMLDivElement = document.createElement("div")
  leftBar: HTMLDivElement = document.createElement("div")
  blocksDiv: HTMLDivElement = document.createElement("div")

  constructor(def: ClauseInput, c: ClauseInstanceInput, owner: Block, clauseIndex: number) {
    super(c.blocks, owner.workspace)
    this.def = def
    this.c = c
    this.owner = owner
    this.clauseIndex = clauseIndex
  }

  getHighlightElement(): HTMLDivElement {
    const element = (this.blocks.length === 0) ? this.div : this.blocks[0].blockDiv
    return element
  }

  draw(container: Block, extraDropDiv: HTMLDivElement | null): HTMLDivElement {
    const acceptor = new ClauseAcceptor(this)

    this.div = document.createElement("div")
    this.div.classList.add("nt-clause")

    if (extraDropDiv !== null) {
      const extraDropzone = interact(extraDropDiv).dropzone({
          accept: ".nt-menu-slot, .nt-block, .nt-cap, .nt-notch"
        , checker: (_1, _2, dropped) => acceptor.checker(dropped)
      })
      extraDropzone.on("drop", () => this.drop() )
      extraDropzone.on("dragenter", () => {
        this.getHighlightElement().classList.add("nt-block-clause-drag-over")
      })
      extraDropzone.on("dragleave", () => {
        this.getHighlightElement().classList.remove("nt-block-clause-drag-over")
      })
    }

    const styleClass = container.getStyleClass()

    this.leftBar = document.createElement("div")
    this.leftBar.classList.add("nt-clause-left-bar")
    this.leftBar.classList.add(`${styleClass}-color`)
    BlockRules.maybeSetColorOverride(container.def.blockColor, this.leftBar)
    this.div.append(this.leftBar)

    this.divider = document.createElement("div")
    this.divider.classList.add("nt-clause-divider")
    this.divider.classList.add(`${styleClass}-color`)
    BlockRules.maybeSetColorOverride(container.def.blockColor, this.divider)
    this.div.append(this.divider)

    const dividerText = StringUtils.toStrNotEmpty(this.def.action, "")
    if (StringUtils.isNotNullOrEmpty(dividerText.trim())) {
      const dividerTextDiv = document.createElement("div")
      dividerTextDiv.classList.add("nt-clause-divider-text")
      dividerTextDiv.innerText = dividerText
      this.divider.append(dividerTextDiv)
    }
    // I would much prefer to just put the arrow in the clause directly, but it gets a little
    // tricky to figure out where to place it veritically with dividers, empty clauses, etc.
    // So this makes this mildly easier, but still a little hacky (see the CSS file for more).
    // -Jeremy B August 2020
    const arrow = Arrow.draw()
    this.divider.append(arrow)

    const dropzone = interact(this.div).dropzone({
        accept: ".nt-menu-slot, .nt-block, .nt-cap, .nt-notch"
      , checker: (_1, _2, dropped) => acceptor.checker(dropped)
    })
    dropzone.on("drop", () => this.drop() )
    dropzone.on("dragenter", () => {
      this.getHighlightElement().classList.add("nt-block-clause-drag-over")
    })
    dropzone.on("dragleave", () => {
      this.getHighlightElement().classList.remove("nt-block-clause-drag-over")
    })

    this.blocksDiv = document.createElement("div")
    this.blocksDiv.classList.add("nt-clause-blocks")

    if (this.blocks.length === 0) {
      this.setEmpty()
      return this.div
    }

    this.div.append(this.blocksDiv)

    for (var i = 0; i < this.blocks.length; i++) {
      const block = this.blocks[i]
      const siblings = this.blocks.slice(i + 1)
      if (!(this.owner.dragData instanceof ActiveDragData) || this.owner.b.instanceId === null) {
        throw new Error("Cannot draw a clause for a block missing drag data.")
      }
      const dragData = new ClauseDragData(block, this.owner.dragData.chainIndex, i, this.owner.b.instanceId, siblings, this.clauseIndex)
      block.draw(dragData)
    }

    BlockCollection.appendBlocks(this.blocksDiv, this.blocks, "nt-block-clause")

    return this.div
  }

  setEmpty(): void {
    this.div.classList.add("nt-clause-empty")
    this.div.append(Notch.drawClause(false, this))

    const dropElement = document.createElement("div")
    dropElement.className = "nt-clause-drop"
    this.div.append(dropElement)

    this.div.append(Notch.drawClause(true,  this))
  }

  setNonEmpty(): void {
    this.div.classList.remove("nt-clause-empty")
  }

  resetOwned(): void {
    for (var i = 0; i < this.blocks.length; i++) {
      const block = this.blocks[i]
      if (!(this.owner.dragData instanceof ActiveDragData) || this.owner.b.instanceId === null) {
        throw new Error("Cannot draw a clause for a block missing drag data or instance ID.")
      }
      block.dragData = new ClauseDragData(block, this.owner.dragData.chainIndex, i, this.owner.b.instanceId, this.blocks.slice(i + 1), this.clauseIndex)
      block.resetOwnedBlocksDragData()
    }
  }

  redrawBlocks(): void {
    this.div.innerHTML = ""
    this.blocksDiv.innerHTML = ""

    this.div.append(this.leftBar)
    this.div.append(this.divider)

    if (this.blocks.length === 0) {
      this.setEmpty()
      return
    }

    this.setNonEmpty()

    this.div.append(this.blocksDiv)

    for (var i = 0; i < this.blocks.length; i++) {
      const block = this.blocks[i]
      if (!(this.owner.dragData instanceof ActiveDragData) || this.owner.b.instanceId === null) {
        throw new Error("Cannot draw a clause for a block missing drag data or instance ID.")
      }
      block.dragData = new ClauseDragData(block, this.owner.dragData.chainIndex, i, this.owner.b.instanceId, this.blocks.slice(i + 1), this.clauseIndex)
      block.resetOwnedBlocksDragData()
    }

    BlockCollection.appendBlocks(this.blocksDiv, this.blocks, "nt-block-clause")
  }

  enableDropZones(): void {
    if (DragManager.isValidDrop(this.owner.workspace.containerId, (dragState) => ClauseAcceptor.isLandingSpot(this, dragState))) {
      this.div.classList.add("nt-allowed-drop")
    }

    for (const block of this.blocks) {
      block.enableDropZones()
    }
  }

  disableDropZones(): void {
    this.div.classList.remove("nt-allowed-drop")

    for (var block of this.blocks) {
      block.disableDropZones()
    }
  }

  drop(): void {
    DragManager.drop( (newBlocks) => {
      this.getHighlightElement().classList.remove("nt-block-clause-drag-over")
      this.insertBlocks(0, newBlocks)
      this.div.classList.remove("nt-clause-empty")

      const changedBlock = newBlocks[0]
      this.owner.workspace.programChanged(new BlockChangedEvent(changedBlock))
    })
  }

}

export { Clause }
